Description: launchpad integration items
Author: ?

Index: nemo-3.3.4/configure.in
===================================================================
--- nemo-3.3.4.orig/configure.in	2012-01-28 21:23:15.730238374 -0500
+++ nemo-3.3.4/configure.in	2012-01-28 21:23:25.054284614 -0500
@@ -318,6 +318,7 @@
 	gio-unix-2.0 >= glib_minver
 	gsettings-desktop-schemas
 	libnotify >= notify_minver
+	launchpad-integration-3.0
 ])
 
 AC_SUBST(NEMO_CFLAGS)
Index: nemo-3.3.4/src/nemo-shell-ui.xml
===================================================================
--- nemo-3.3.4.orig/src/nemo-shell-ui.xml	2012-01-28 21:23:15.742238447 -0500
+++ nemo-3.3.4/src/nemo-shell-ui.xml	2012-01-28 21:23:25.054284614 -0500
@@ -113,6 +113,7 @@
 		<menuitem action="NemoHelpSort"/>
 		<menuitem action="NemoHelpLost"/>
 		<menuitem action="NemoHelpShare"/>
+		<placeholder name='LaunchpadItems'/>
 		<separator/>
 		<menuitem name="About Nemo" action="About Nemo"/>
 	</menu>
Index: nemo-3.3.4/src/nemo-window-menus.c
===================================================================
--- nemo-3.3.4.orig/src/nemo-window-menus.c	2012-01-28 21:23:15.758238525 -0500
+++ nemo-3.3.4/src/nemo-window-menus.c	2012-01-28 21:25:43.062968956 -0500
@@ -60,6 +60,7 @@
 #include <libnemo-private/nemo-signaller.h>
 #include <libnemo-private/nemo-trash-monitor.h>
 #include <string.h>
+#include <launchpad-integration.h>
 
 #define MENU_PATH_EXTENSION_ACTIONS                     "/MenuBar/File/Extension Actions"
 #define POPUP_PATH_EXTENSION_ACTIONS                     "/background/Before Zoom Items/Extension Actions"
@@ -1297,6 +1298,8 @@
 	/* add the UI */
 	gtk_ui_manager_add_ui_from_resource (ui_manager, "/org/gnome/nemo/nemo-shell-ui.xml", NULL);
 
+	launchpad_integration_add_ui (ui_manager, "/MenuBar/Help/LaunchpadItems");
+
 	nemo_window_initialize_trash_icon_monitor (window);
 	nemo_window_initialize_go_menu (window);
 }
