Index: nemo-3.3.92/po/POTFILES.in
===================================================================
--- nemo-3.3.92.orig/po/POTFILES.in	2012-02-06 20:44:45.000000000 +0100
+++ nemo-3.3.92/po/POTFILES.in	2012-03-20 11:19:38.107758604 +0100
@@ -4,6 +4,7 @@
 data/nemo.xml.in
 data/nemo-autorun-software.desktop.in.in
 data/nemo.desktop.in.in
+debian/mount-archive.desktop.in
 eel/eel-canvas.c
 eel/eel-editable-label.c
 eel/eel-glib-extensions.c
