Index: nemo-3.3.92/src/nemo-progress-ui-handler.c
===================================================================
--- nemo-3.3.92.orig/src/nemo-progress-ui-handler.c	2012-02-06 20:44:45.000000000 +0100
+++ nemo-3.3.92/src/nemo-progress-ui-handler.c	2012-03-20 11:19:38.123758605 +0100
@@ -201,6 +201,7 @@
 progress_ui_handler_ensure_window (NemoProgressUIHandler *self)
 {
 	GtkWidget *vbox, *progress_window;
+	const gchar *desktop_environment = g_getenv ("DESKTOP_SESSION");
 	
 	if (self->priv->progress_window != NULL) {
 		return;
@@ -220,6 +221,10 @@
 				 GTK_WIN_POS_CENTER);
 	gtk_window_set_icon_name (GTK_WINDOW (progress_window),
 				"system-file-manager");
+	if ((!g_strcmp0(desktop_environment, "ubuntu")) ||
+	       (!g_strcmp0(desktop_environment, "ubuntu-2d")))
+	    gtk_window_set_skip_taskbar_hint (GTK_WINDOW (progress_window),
+				TRUE);
 
 	vbox = gtk_box_new (GTK_ORIENTATION_VERTICAL, 0);
 	gtk_box_set_spacing (GTK_BOX (vbox), 5);
